s3/s3_transport/include/s3_transport.hpp

```diff
@@ -1067,58 +1067,90 @@ namespace irods::experimental::io::s3_transport
                 ? config_.number_of_cache_transfer_threads
                 : cache_file_size / minimum_part_size == 0 ? 1 : cache_file_size / minimum_part_size;
 
-            // Calculate the number of parts.  This is usually a one-to-one mapping to number of threads but if
-            // parts per thread > config_.max_single_part_upload_size then need to break it into more parts
-            unsigned int number_of_parts = config_.number_of_cache_transfer_threads;
-            if (cache_file_size > number_of_parts * config_.max_single_part_upload_size) {
-                number_of_parts
-                  = cache_file_size % config_.max_single_part_upload_size == 0
-                  ? cache_file_size / config_.max_single_part_upload_size
-                  : cache_file_size / config_.max_single_part_upload_size + 1;
-            }
+        //int64_t preferred_part_size = config_.max_single_part_upload_size;
+        // Preferred part size is the largest part size that will be attempted. At 1 GiB, that still allows the largest possible
+        // file size (1 TiB) to be uploaded within the 10,000 part limit imposed by AWS.
+        int64_t preferred_part_size = 1LL*1024*1024*1024;
+        bool break_out = false;
+
+        while (!break_out) {
+
+                // Calculate the number of parts.
+                unsigned int number_of_parts = config_.number_of_cache_transfer_threads;
+                if (cache_file_size > number_of_parts * preferred_part_size) {
+                    number_of_parts
+                      = cache_file_size % preferred_part_size == 0
+                      ? cache_file_size / preferred_part_size
+                      : cache_file_size / preferred_part_size + 1;
+                }
```

Comment:
Is it possible for this to change during the loop?

If not, consider moving it out.

Reply:
The preferred_part_size can change which means the number_of_parts can.